README.md
# Route 53 Domain Manager

CLI tool that will iterate through all of the domains you have registered in
Route 53 and update the Registrant, Administrative and Technical contact
information.

[![CircleCI](https://circleci.com/gh/cachelab/r53_domain_manager.svg?style=svg)](https://circleci.com/gh/cachelab/r53_domain_manager)

## Usage

This cli is configured by the following parameters:

### Examples

```bash
r53dm list

# Parameter is required.
r53dm describe \
--domain=cachelab.co

# All parameters are required and will be updated accordingly.
r53dm update \
--domain=cachelab.co \
--addressline1="1234 Jones Drive" \
--city=Littleton \
--state=CO \
--zip=80127 \
--email=hello@cachelab.co \
--first=Andrew \
--last=Puch \
--phone=+1.1231231234
```

#### Parameters

```bash
domain       Domain you wish to update.
addressline1 Address line 1 you want to change.
city         City you want to change.
state        State you want to change.
zip          Zip you want to change.
email        Email you want to change.
first        First name you want to change.
last         Last name you want to change.
phone        Phone number you want to change.
```

## Contributing

* `make run` - runs the r53dm cli in a docker container
* `make build` - builds your r53dm docker container
* `make vet` - go fmt and vet code
* `make test` - run unit tests

Before you submit a pull request please update the semantic version inside of
`service.go` with what you feel is appropriate and then edit the `CHANGELOG.md`
with your changes and follow a similar structure to what is there.


CHANGELOG.md
## 1.0.1 (November 19 2019)

* Add LICENSE and start unit tests.
* README updates.

## 1.0.0 (November 18 2019)

* Initial release.


commands/list_test.go
package commands

import (
	"testing"

	"github.com/mitchellh/cli"
)

func TestListCommand(t *testing.T) {
	var _ cli.Command = &ListCommand{}
}


commands/update_test.go
package commands

import (
	"testing"

	"github.com/mitchellh/cli"
)

func TestUpdateCommand(t *testing.T) {
	var _ cli.Command = &UpdateCommand{}
}


commands/describe_test.go
package commands

import (
	"testing"

	"github.com/mitchellh/cli"
)

func TestDescribeCommand(t *testing.T) {
	var _ cli.Command = &DescribeCommand{}
}


service.go
package main

import (
	"fmt"
	"os"

	"r53_domain_manager/commands"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/route53domains"
	"github.com/mitchellh/cli"
)

const version = "1.0.1"

type Service struct{}

func (svc *Service) init() {
	client := route53domains.New(session.New())

	ui := &cli.PrefixedUi{
		AskPrefix:    "",
		OutputPrefix: "",
		InfoPrefix:   "",
		ErrorPrefix:  "",
		Ui: &cli.BasicUi{
			Writer: os.Stdout,
			Reader: os.Stdin,
		},
	}

	c := cli.NewCLI("app", version)
	c.Args = os.Args[1:]
	c.Commands = map[string]cli.CommandFactory{
		"list": func() (cli.Command, error) {
			return &commands.ListCommand{
				Ui:     ui,
				Client: client,
			}, nil
		},
		"describe": func() (cli.Command, error) {
			return &commands.DescribeCommand{
				Ui:     ui,
				Client: client,
			}, nil
		},
		"update": func() (cli.Command, error) {
			return &commands.UpdateCommand{
				Ui:     ui,
				Client: client,
			}, nil
		},
	}

	exitStatus, err := c.Run()
	if err != nil {
		fmt.Println(err)
	}

	os.Exit(exitStatus)
}


commands/list.go
package commands

import (
	"strconv"
	"strings"

	"github.com/mitchellh/cli"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/route53domains"
	"github.com/aws/aws-sdk-go/service/route53domains/route53domainsiface"
	"github.com/olekukonko/tablewriter"
)

type ListCommand struct {
	Ui     cli.Ui
	Client route53domainsiface.Route53DomainsAPI
}

func (c *ListCommand) Help() string {
	helpText := `
Usage: r53dm list

  This will list all the domains you have purchased using Route 53.
`

	return strings.TrimSpace(helpText)
}

func (c *ListCommand) Synopsis() string {
	return "Lists domains that have purchased"
}

func (c *ListCommand) Run(args []string) int {
	var marker *string
	var allDomains []*route53domains.DomainSummary

	more := true

	for more {
		domains, err := c.Client.ListDomains(&route53domains.ListDomainsInput{
			MaxItems: aws.Int64(20),
			Marker:   marker,
		})
		if err != nil {
			c.Ui.Output(err.Error())
			return 1
		}

		allDomains = append(allDomains, domains.Domains...)

		if domains.NextPageMarker == nil {
			more = false
		} else {
			marker = domains.NextPageMarker
		}
	}

	tableString := &strings.Builder{}
	table := tablewriter.NewWriter(tableString)
	table.SetHeader([]string{"Domain", "Auto Renew", "Transfer Lock", "Expiry"})

	for _, domain := range allDomains {
		table.Append([]string{
			*domain.DomainName,
			strconv.FormatBool(*domain.AutoRenew),
			strconv.FormatBool(*domain.TransferLock),
			domain.Expiry.Format("2006-01-02"),
		})
	}

	table.Render()

	c.Ui.Output(tableString.String())

	return 0
}


commands/describe.go
package commands

import (
	"flag"
	"strings"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/route53domains"
	"github.com/aws/aws-sdk-go/service/route53domains/route53domainsiface"
	"github.com/mitchellh/cli"
	"github.com/olekukonko/tablewriter"
)

type DescribeCommand struct {
	Ui     cli.Ui
	Client route53domainsiface.Route53DomainsAPI
	Domain string
}

func (c *DescribeCommand) Help() string {
	helpText := `
Usage: r53dm describe [options]

  This will get more detailed information on the domain according to what
  you pass in as options.

Options:

  --domain=example.com   The domain you wish to get more information for.
`

	return strings.TrimSpace(helpText)
}

func (c *DescribeCommand) Synopsis() string {
	return "Describes domain passed in"
}

func (c *DescribeCommand) Run(args []string) int {
	cmdFlags := flag.NewFlagSet("agent", flag.ContinueOnError)
	cmdFlags.StringVar(&c.Domain, "domain", "example.com", "The domain you want to describe")
	cmdFlags.Usage = func() { c.Ui.Output(c.Help()) }

	if err := cmdFlags.Parse(args); err != nil {
		c.Ui.Output(err.Error())
		return 1
	}

	if c.Domain == "example.com" {
		c.Ui.Output("\nInvalid domain\n")
		cmdFlags.Usage()
		return 1
	}

	domain, err := c.Client.GetDomainDetail(&route53domains.GetDomainDetailInput{
		DomainName: aws.String(c.Domain),
	})
	if err != nil {
		c.Ui.Output(err.Error())
		return 1
	}

	tableString := &strings.Builder{}
	table := tablewriter.NewWriter(tableString)
	table.SetHeader([]string{"Contact Type", "Address Line 1", "City", "State", "Zip", "Email", "First", "Last", "Phone"})

	table.Append([]string{
		"Admin",
		*domain.AdminContact.AddressLine1,
		*domain.AdminContact.City,
		*domain.AdminContact.State,
		*domain.AdminContact.ZipCode,
		*domain.AdminContact.Email,
		*domain.AdminContact.FirstName,
		*domain.AdminContact.LastName,
		*domain.AdminContact.PhoneNumber,
	})

	table.Append([]string{
		"Registrant",
		*domain.RegistrantContact.AddressLine1,
		*domain.RegistrantContact.City,
		*domain.RegistrantContact.State,
		*domain.RegistrantContact.ZipCode,
		*domain.RegistrantContact.Email,
		*domain.RegistrantContact.FirstName,
		*domain.RegistrantContact.LastName,
		*domain.RegistrantContact.PhoneNumber,
	})

	table.Append([]string{
		"Technical",
		*domain.TechContact.AddressLine1,
		*domain.TechContact.City,
		*domain.TechContact.State,
		*domain.TechContact.ZipCode,
		*domain.TechContact.Email,
		*domain.TechContact.FirstName,
		*domain.TechContact.LastName,
		*domain.TechContact.PhoneNumber,
	})

	table.Render()

	c.Ui.Output(tableString.String())

	return 0
}


commands/update.go
package commands

import (
	"flag"
	"strings"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/route53domains"
	"github.com/aws/aws-sdk-go/service/route53domains/route53domainsiface"
	"github.com/mitchellh/cli"
)

type UpdateCommand struct {
	Ui           cli.Ui
	Client       route53domainsiface.Route53DomainsAPI
	Domain       string
	AddressLine1 string
	City         string
	State        string
	Zip          string
	Email        string
	First        string
	Last         string
	Phone        string
}

func (c *UpdateCommand) Help() string {
	helpText := `
Usage: r53dm update [options]

  This will update the contact information on the domain according to what
  you pass in as options.

Options:

  --domain=example.com               Domain you wish to update.
  --addressline1="1234 Jones Drive"  Address line 1 you want to change.
  --city=Littleton                   City you want to change.
  --state=CO                         State you want to change.
  --zip=80127                        Zip you want to change.
  --email=hello@cachelab.co          Email you want to change.
  --first=Andrew                     First name you want to change.
  --last=Puch                        Last name you want to change.
  --phone=+1.1231231234              Phone number you want to change.
`

	return strings.TrimSpace(helpText)
}

func (c *UpdateCommand) Synopsis() string {
	return "Updates domain contact information"
}

func (c *UpdateCommand) Run(args []string) int {
	cmdFlags := flag.NewFlagSet("agent", flag.ContinueOnError)
	cmdFlags.StringVar(&c.Domain, "domain", "example.com", "The domain you want to update")
	cmdFlags.StringVar(&c.AddressLine1, "addressline1", "1234 Jones Drive", "The address line 1 you want to change")
	cmdFlags.StringVar(&c.City, "city", "CacheLab", "The city you want to change")
	cmdFlags.StringVar(&c.State, "state", "CacheLab", "The state you want to change")
	cmdFlags.StringVar(&c.Zip, "zip", "12345", "The zipcode you want to change")
	cmdFlags.StringVar(&c.Email, "email", "hello@example.com", "The email you want to change")
	cmdFlags.StringVar(&c.First, "first", "CacheLab", "The first name you want to change")
	cmdFlags.StringVar(&c.Last, "last", "CacheLab", "The last name you want to change")
	cmdFlags.StringVar(&c.Phone, "phone", "+1.1231231234", "The phone number you want to change")
	cmdFlags.Usage = func() { c.Ui.Output(c.Help()) }

	if err := cmdFlags.Parse(args); err != nil {
		c.Ui.Output(err.Error())
		return 1
	}

	if c.Domain == "example.com" {
		c.Ui.Output("\nInvalid domain\n")
		cmdFlags.Usage()
		return 1
	}

	if c.AddressLine1 == "1234 Jones Drive" {
		c.Ui.Output("\nInvalid address line 1\n")
		cmdFlags.Usage()
		return 1
	}

	if c.City == "CacheLab" {
		c.Ui.Output("\nInvalid city\n")
		cmdFlags.Usage()
		return 1
	}

	if c.State == "CacheLab" {
		c.Ui.Output("\nInvalid state\n")
		cmdFlags.Usage()
		return 1
	}

	if c.Zip == "12345" {
		c.Ui.Output("\nInvalid zip\n")
		cmdFlags.Usage()
		return 1
	}

	if c.Email == "hello@example.com" {
		c.Ui.Output("\nInvalid email\n")
		cmdFlags.Usage()
		return 1
	}

	if c.First == "CacheLab" {
		c.Ui.Output("\nInvalid first\n")
		cmdFlags.Usage()
		return 1
	}

	if c.Last == "CacheLab" {
		c.Ui.Output("\nInvalid last\n")
		cmdFlags.Usage()
		return 1
	}

	if c.Phone == "+1.1231231234" {
		c.Ui.Output("\nInvalid phone\n")
		cmdFlags.Usage()
		return 1
	}

	contactDetail := &route53domains.ContactDetail{
		AddressLine1: aws.String(c.AddressLine1),
		City:         aws.String(c.City),
		Email:        aws.String(c.Email),
		FirstName:    aws.String(c.First),
		LastName:     aws.String(c.Last),
		PhoneNumber:  aws.String(c.Phone),
		State:        aws.String(c.State),
		ZipCode:      aws.String(c.Zip),
	}

	_, err := c.Client.UpdateDomainContact(&route53domains.UpdateDomainContactInput{
		DomainName:        aws.String(c.Domain),
		AdminContact:      contactDetail,
		RegistrantContact: contactDetail,
		TechContact:       contactDetail,
	})
	if err != nil {
		c.Ui.Output(err.Error())
		return 1
	}

	c.Ui.Output("Domain updated")

	return 0
}
